selection.go

```diff
@@ -0,0 +1,231 @@
+package ice
+
+import (
+	"net"
+	"sort"
+	"time"
+
+	"github.com/pion/logging"
+	"github.com/pion/stun"
+)
+
+type PairCandidateSelector interface {
+	Start()
+	ContactCandidates()
+	PingCandidate(local, remote *Candidate)
+	HandleSucessResponse(m *stun.Message, local, remote *Candidate, remoteAddr net.Addr)
+	HandleBindingRequest(m *stun.Message, local, remote *Candidate)
+}
+
+type ControllingSelector struct {
+	agent         *Agent
+	nominatedPair *candidatePair
+	log           logging.LeveledLogger
+}
+
+func (s *ControllingSelector) Start() {
+	go func() {
```

Comment:
Nice, this all looks really good!

I think the last thing is we can drop this part (now that `nominatePair` is called in `ContactCandidates`) and then move the `ConnectionStateFailed` into a timer that gets reset when `updateConnectionState` == `Checking` and if we are stuck in `Checking` for more then n seconds we go to failed?

Reply:
Sorry the code was a bit confusing, I updated it. We still use this to actually select the nominatedPair.

So we need two things:

- a way to nominate a pair, I'm using a timeout in here, but we could use another criteria. I know you suggested to nominate the first pair we get, but I don't really like that too much.. I think this way (which is not ideal either) at least has the advantage of selecting by priority.
- the timeout you mentioned. We are already doing that in here, we could check if we are stuck in checking instead, but not sure if that's is actually better, just because there are other reasons to be stuck on a checking state (like if we don't get an aswer from the signaling system), I feel a timeout to wait for candidates may be easier to understand from the point of view of the user of the library.